### DISCUSSION TITLE
tutorials/plugins/editor/import_plugins

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#147] [User-contributed notes]
# tutorials/plugins/editor/import_plugins

Introduction: An import plugin is a special type of editor tool that allows custom resources to be imported by Godot and be treated as first-class resources. The editor itself comes bundled with a ...

https://docs.godotengine.org/en/stable/tutorials/plugins/editor/import_plugins.html

**Comment by horseyhorsey** [comment]
`Unimplemented _get_priority in add-on.`
`BUG: File queued for import, but can't be imported, importer for type '' not found.`

Add:
```
func _get_priority():
	return 2
```

**Reply by scriptsengineer** [reply]
Why 2?

**Reply by Jeremi360** [reply]
I try to import some files as JSON.
I added :
```gdscript
func _get_priority():
	return 2

# after that I also need to add this
func _get_import_order() -> int:
	return ResourceImporter.ImportOrder.IMPORT_ORDER_DEFAULT
```
But now I getting this errors:
```
  editor/import/editor_import_plugin.cpp:123 - Condition "!d.has_all(needed)" is true.
  core/io/file_access.cpp:812 - Condition "f.is_null()" is true. Continuing.
```

**Comment by DissonantVoid** [comment]
Can this be used to override a built-in importer while still relying on its functionality? for example can I override the gltf importer so all my meshes automatically use a toon material for a stylized game, while using import functionality from the built-in gltf importer so I don't have to rewrite one from scratch?

**Reply by DissonantVoid** [reply]
I can! by either using a script that inherites EditorScenePostImport and attaching it to each gltf file in the import settings, or use a EditorScenePostImportPlugin plugin to automate it further without having to open each file. Hope this helps someone.

**Reply by DissonantVoid** [reply]
Here's the plugin I wrote to automatically apply a toon shader to all meshes on import which is something I struggled for a while to figure out since I couldn't find any guides. make sure to create a new plugin and call `add_scene_post_import_plugin()` there to add your importer script, this is my importer script which can easily be repurposed if needed:

```gdscript
@tool
extends EditorScenePostImportPlugin

var target_material : ShaderMaterial  = preload("res://resources/shaders/toon_material_template.tres")

func _pre_process(scene: Node) -> void:
	iterate(scene)

func iterate(node: Node) -> void:
	# ImporterMeshInstance3D is used instead of MeshInstance3D while importing
	if node is ImporterMeshInstance3D:
		var mesh : ImporterMesh = node.mesh
		
		# replace materials with toon material
		for index in mesh.get_surface_count():
			var material : StandardMaterial3D = mesh.get_surface_material(index)
			var new_material : ShaderMaterial = target_material.duplicate(false) # duplicate(true) will also duplicate the shader which I don't want
			
			# I know that my models will only use an albedo texture (no other pbr texture)
			# and my toon shader only requires an albedo so I only have
			# to pass that
			new_material.set_shader_parameter("albedo_texture", material.albedo_texture)
			
			mesh.set_surface_material(index, new_material)
	
	for child in node.get_children():
		iterate(child)

```
